### PR TITLE
Adds aggregation profiling to the profile API

### PR DESCRIPTION
**Pull request by colings86**
This change refactors some of the query profiling classes to extract common functionality that can be used to profile other features and then adds profiling of aggregations.

Caveats:
- This is very early stage and definitely contains bugs
- No tests exist yet for agg profiling
- The reduce phase is not profiled yet and will always appear as 0. This is because we currently have no infrastructure on the coordinating node for profiling. This will probably be added in a later PR.

Relates to #10538

### CODE REVIEW

**Comment by colings86** [comment]
@jpountz @polyfractal I've made some progress on this and it would be great to get your feedback on how close to ready you think it is.

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java [hunk 1]
```diff
@@ -129,7 +129,8 @@ public boolean needsScores() {
     @Override
     public final LeafBucketCollector getLeafCollector(LeafReaderContext ctx) throws IOException {
         final LeafBucketCollector sub = collectableSubAggregators.getLeafCollector(ctx);
-        return getLeafCollector(ctx, sub);
+        LeafBucketCollector leafCollector = getLeafCollector(ctx, sub);
+        return leafCollector;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java]: let's revert this change?

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactory.java [hunk 2]
```diff
@@ -28,13 +28,139 @@
...
+                Scorer scorer;
+
+                @Override
+                public void setScorer(Scorer scorer) throws IOException {
+                    this.scorer = scorer;
+                }
+
+                @Override
+                public void collect(int doc, long bucket) throws IOException {
+                    aggregators = bigArrays.grow(aggregators, bucket + 1);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactory.java]: should we move it under the `if collector == null` check to save some cycles in the common case?

File: core/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    protected ArrayList<E> elements;
+    /** A list of top-level "roots".  Each root can have its own tree of profiles */
+    protected ArrayList<Integer> roots;
+    /** A temporary stack used to record where we are in the dependency tree.  Only used by scoring queries */
+    protected Deque<Integer> stack;
+    private int currentToken = 0;
+
+    public AbstractInternalProfileTree() {
+        timings = new ArrayList<>(10);
+        stack = new LinkedBlockingDeque<>(10);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java]: hmm why do we need a blocking deque? a regular array deque should be enough?

File: core/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+    protected ArrayList<PB> timings;
+    /** Maps the Query to it's list of children.  This is basically the dependency tree */
+    protected ArrayList<ArrayList<Integer>> tree;
+    /** A list of the original queries, keyed by index position */
+    protected ArrayList<E> elements;
+    /** A list of top-level "roots".  Each root can have its own tree of profiles */
+    protected ArrayList<Integer> roots;
+    /** A temporary stack used to record where we are in the dependency tree.  Only used by scoring queries */
+    protected Deque<Integer> stack;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java]: then could it be only in the query impl?

File: core/src/main/java/org/elasticsearch/search/profile/aggregation/AggregationTimingType.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.profile.aggregation;
+
+import java.util.Locale;
+
+public enum AggregationTimingType {
+    INITIALISE,
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/profile/aggregation/AggregationTimingType.java]: nit-pick but I'm wondering that we might be using the american -ize in other parts of the response so we might want to be consistent?

File: core/src/main/java/org/elasticsearch/search/profile/aggregation/ProfilingAggregator.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+
+import java.io.IOException;
+
+public class ProfilingAggregator extends Aggregator {
+
+    private Aggregator delegate;
+    private AggregationProfileBreakdown profileBreakdown;
+    private AggregationProfiler profiler;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/profile/aggregation/ProfilingAggregator.java]: can all fields be final?

**Comment by jpountz** [comment]
I like that this change is not too intrusive, the only hack I could spot is the getWrappedClass on the multi-bucket aggregator, which I think is fine. We might want to check whether we can make the profile wrappers pkg-private. Before giving an LGTM I would just like to do another iteration to better understand how the profile tree is built, but in general this looks on the right path to me.

File: core/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java (hunk 4 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java]: The comment was not updated after the refactoring to make this class abstract. This stack is used to record the position in the dependency tree and is needed for both the query and agg impls

File: core/src/main/java/org/elasticsearch/search/profile/aggregation/ProfilingAggregator.java (hunk 6 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/profile/aggregation/ProfilingAggregator.java]: dleegate and profiler can be final but profileBreakdown is set in `preCollection()` so cannot be final.

**Comment by colings86** [comment]
@jpountz thanks for the review. I've updated this PR with your comments. Note that the profile tree is the same logic as used for the query profiling.

**Comment by jpountz** [comment]
LGTM
